mfa/push/src/main/kotlin/com/pingidentity/mfa/push/PushUriParser.kt

```diff
@@ -0,0 +1,262 @@
+     */
+    @JvmStatic
+    fun format(credential: PushCredential): String {
+        // Format the label part in a way that preserves the colon character
+        val encodedLabel = if (credential.issuer.isNotEmpty()) {
+            val encodedIssuer = Uri.encode(credential.issuer)
+            val encodedAccount = Uri.encode(credential.accountName)
+            "$encodedIssuer:$encodedAccount"
+        } else {
+            Uri.encode(credential.accountName)
+        }
+
+        // Create registration and authentication endpoints
+        val regEndpoint = "${credential.serverEndpoint}?_action=register"
+        val authEndpoint = "${credential.serverEndpoint}?_action=authenticate"
+
+        // Encode the endpoints in Base64
+        val encodedRegEndpoint = encodeBase64(regEndpoint)
+        val encodedAuthEndpoint = encodeBase64(authEndpoint)
+
+        val uriBuilder = StringBuilder()
+            .append("$PUSHAUTH_SCHEME://push/")
+            .append(encodedLabel)
+            .append("?$REG_ENDPOINT_PARAM=").append(Uri.encode(encodedRegEndpoint))
+            .append("&$AUTH_ENDPOINT_PARAM=").append(Uri.encode(encodedAuthEndpoint))
+            .append("&$SHARED_SECRET_PARAM=").append(recodeBase64NoWrapValueToUrlSafeNoWrap(credential.sharedSecret))
+
+        if (credential.issuer.isNotEmpty()) {
+            uriBuilder.append("&$ISSUER_PARAM=").append(Uri.encode(credential.issuer))
+        }
+
+        // Add optional parameters if present
+        if (credential.userId?.isNotEmpty() == true) {
+            val encodedUserId = encodeBase64(credential.userId)
+            uriBuilder.append("&$USER_ID_PARAM=").append(Uri.encode(encodedUserId))
+        }
+
+        if (credential.resourceId.isNotEmpty()) {
+            val encodedResourceId = encodeBase64(credential.resourceId)
+            uriBuilder.append("&$PUSH_RESOURCE_ID_PARAM=").append(Uri.encode(encodedResourceId))
+        }
+
+        // Add the new fields if present
+        if (!credential.imageURL.isNullOrEmpty()) {
+            val encodedImageURL = encodeBase64(credential.imageURL)
+            uriBuilder.append("&$IMAGE_URL_PARAM=").append(Uri.encode(encodedImageURL))
+        }
+
+        // Format background color (removing # if present) when adding to URI
+        if (!credential.backgroundColor.isNullOrEmpty()) {
+            val colorValue = formatBackgroundColor(credential.backgroundColor)
+            uriBuilder.append("&$BACKGROUND_COLOR_PARAM=").append(Uri.encode(colorValue))
+        }
+
+        // Add policies if present, encoding them in Base64 if they're JSON
+        if (!credential.policies.isNullOrEmpty()) {
+            val encodedPolicies = encodeBase64(credential.policies)
+            uriBuilder.append("&$POLICIES_PARAM=").append(Uri.encode(encodedPolicies))
+        }
+
+        return uriBuilder.toString()
+    }
+
+    @JvmStatic
+    fun registrationParameters(uri: String): Map<String, String> {
+        val parsedUri = Uri.parse(uri)
```

Comment:
May use uri.toUri()

Reply:
weird, it was not working for non `http://` scheme before. Now worked fine.